File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] answers = new int[nums.length];
        int total = 1;
        int numZeroes = 0;
        for (int num : nums) {
            if (num != 0) {
               total *= num; 
            } else {
                numZeroes++;
            }
        }

        if (numZeroes > 1) {
            return answers;
        }

        for (int i = 0; i < nums.length; ++i) {
            if (numZeroes == 0) {
                answers[i] = total / nums[i];
            } else {
                if (nums[i] == 0) {
                    answers[i] = total;
                } else {
                    answers[i] = 0;
                }
            }
        }
        return answers;
    }
}  

File: Data Structures & Algorithms/valid-sudoku/submission-0.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        blocks = defaultdict(list)
        cols = defaultdict(list)
        
        
        for r in range(9):
            for c in range(9):
                curr_val = board[r][c]
                cols[c].append(curr_val)
                blocks[(r // 3, c // 3)].append(curr_val)

        for row in board:
            if self.hasDupe(row):
                return False

        for col in cols.values():
            if self.hasDupe(col):
                return False
        
        print(blocks)
        for block in blocks.values():
            if self.hasDupe(block):
                return False
        return True
    
    def hasDupe(self, nums: List[str]) -> bool:
        seen = set()
        for num in nums:
            if num == ".":
                continue
            if num in seen:
                return True
            seen.add(num)
        return False